src/Avalonia.Base/Layout/LayoutManager.cs

```diff
@@ -362,7 +362,7 @@ private bool RaiseEffectiveViewportChanged()
                         if (viewport != l.Viewport)
                         {
                             l.Listener.EffectiveViewportChanged(new EffectiveViewportChangedEventArgs(viewport));
-                            _effectiveViewportChangedListeners[i] = new EffectiveViewportChangedListener(l.Listener, viewport);
+                            l.Viewport = viewport;
```

Comment:
I might be wrong, but it seems like previously shared array was modified and reused later on next RaiseEffectiveViewportChanged call.
But now copy of the array is modified. And structures in original _effectiveViewportChangedListeners are remaining the same.

I don't know if it's a critical problem, nor I don't know why ArrayPool was in this code before unused. Just suspicious.
@grokys

Reply:
Tests are failing, might be related

Reply:
Good catch @maxkatz6, I missed it is struct. Then it makes even less sense to make a copy of structs and not use it for anything.
I changed it to class so we are copying only references and it makes more sense now to me. I don't think memory concern is important in this situation.
Unit tests were passing on my machine, I think there is something else that failed (it was passing even with the struct).
I added another unit test that shows the case that was failing before - tooltips that were closed during this call.

Reply:
Yeah this looks wrong. As @maxkatz6 says, the `_effectiveViewportChangedListeners` was being modified before, now it's only the copy that's being modified. I'll take a look and try to demonstrate the problem with a failing test.

As for why the `ArrayPool` was never used before, I suspect that is unintentional.

Reply:
Ah no, I need to wake up this morning it seems. This looks correct to me.